lib/Dialect/FIRRTL/FFI/FIRRTLFFIContext.cpp

```diff
@@ -0,0 +1,1683 @@
+
+MLIRContext *ModuleContext::getContext() const { return ffiCtx.mlirCtx.get(); }
+
+InFlightDiagnostic ModuleContext::emitError(const Twine &message) {
+  return emitError(ffiCtx.mockSMLoc(), message);
+}
+InFlightDiagnostic ModuleContext::emitError(SMLoc loc, const Twine &message) {
+  ffiCtx.emitError(message.str());
+  auto err = mlir::emitError(translateLocation(loc), message);
+  err.abandon();
+  return err;
+}
+
+Location ModuleContext::translateLocation(llvm::SMLoc loc) {
+  (void)loc;
+  return ffiCtx.mockLoc();
+}
+
+//
+// WhenContext
+//
+
+WhenContext::WhenContext(ModuleContext &moduleCtx, firrtl::WhenOp whenOp)
+    : moduleCtx{moduleCtx}, whenOp{whenOp} {
+  newScope();
+}
+
+bool WhenContext::hasElseRegion() { return whenOp.hasElseRegion(); }
+
+void WhenContext::createElseRegion() {
+  whenOp.createElseRegion();
+  newScope();
+}
+
+Block &WhenContext::currentBlock() {
+  if (!whenOp.hasElseRegion()) {
+    return whenOp.getThenBlock();
+  }
+  return whenOp.getElseBlock();
+}
+
+void WhenContext::newScope() {
+  scope.reset();
+  scope.emplace(
+      // Declarations within the suite are scoped to within the suite.
+      moduleCtx, &currentBlock(),
+      // After parsing the when region, we can release any new entries
+      // in unbundledValues since the symbol table entries that refer
+      // to them will be gone.
+      moduleCtx.unbundledValues);
+}
+
+std::optional<std::pair<StringRef, StringRef>>
+parseUntilUnbalanced(StringRef input, char openingBracket,
+                     char closingBracket) {
+  assert(!input.empty() && input[0] == openingBracket);
+
+  size_t balance = 0;
+  for (size_t i = 0; i < input.size(); i++) {
+    if (input[i] == openingBracket) {
+      balance++;
+    } else if (input[i] == closingBracket) {
+      balance--;
+      if (balance == 0) {
+        return std::make_pair(input.substr(1, i - 1), input.substr(i + 1));
+      }
+    }
+  }
+  return std::nullopt;
+}
+
+} // namespace circt::chirrtl::details
+
+FFIContext::FFIContext() : mlirCtx{std::make_unique<MLIRContext>()} {
+  mlirCtx->loadDialect<CHIRRTLDialect>();
+  mlirCtx->loadDialect<FIRRTLDialect, hw::HWDialect>();
+
+  module = std::make_unique<mlir::ModuleOp>(mlir::ModuleOp::create(mockLoc()));
+  opBuilder = std::make_unique<mlir::OpBuilder>(module->getBodyRegion());
+}
+
+void FFIContext::setErrorHandler(
+    std::function<void(std::string_view message)> handler) {
+  errorHandler = std::move(handler);
+}
+
+void FFIContext::emitError(std::string_view message, bool recoverable) const {
+  if (errorHandler) {
+    errorHandler(message);
+  }
+  // TODO: handle the `recoverable` parameter
+}
+
+void FFIContext::visitCircuit(StringRef name) {
+  moduleContext.underlying.reset();
+
+  circuitOp = opBuilder->create<CircuitOp>(mockLoc(), stringRefToAttr(name));
+  circuitTarget = ("~" + name).str();
+}
+
+void FFIContext::visitModule(StringRef name) {
+  RA_EXPECT(auto &circuitOp, this->circuitOp);
```

Comment:
None of these return any useful indication that there is a failure.  I don't think RA_EXPECT really is pulling its weight either.  But given that there shouldn't be this context in the first place, these problems would go away.